api/api/routers/mcp/_mcp_observability_middleware.py

```diff
@@ -0,0 +1,341 @@
+from core.domain.tenant_data import TenantData
+from core.utils.background import add_background_task
+from core.utils.json_utils import extract_json_str
+
+from ._mcp_observer_agent import mcp_tool_call_observer_agent
+
+logger = logging.getLogger(__name__)
+
+# Track tool call start times per request
+_tool_call_start_times: dict[str, datetime] = {}
+
+
+class MCPObservabilityMiddleware(BaseHTTPMiddleware):
+    """Enhanced middleware to log MCP tool calls and track session observability"""
+
+    def _extract_session_id(self, request: Request) -> str | None:
+        """Extract session ID from request headers"""
+        # Check for MCP session ID header
+        session_id = request.headers.get("mcp-session-id")
+        if session_id:
+            return session_id
+
+        # Check for X-Session-ID header (custom header)
+        session_id = request.headers.get("x-session-id")
+        if session_id:
+            return session_id
+
+        # Return None if no session ID found
+        return None
+
+    async def _get_or_create_session(self, session_id: str | None, tenant: TenantData) -> SessionState:
+        """Get existing session or create a new one using Redis"""
+        session_state, _ = await SessionState.get_or_create(session_id, tenant)
+        return session_state
+
+    def _is_mcp_tool_call(self, request_data: dict[str, Any] | None) -> bool:
+        """Check if this is an MCP tool call request"""
+        if not request_data:
+            return False
+        return request_data.get("method") == "tools/call"
+
+    def _extract_tool_call_info(self, request_data: dict[str, Any]) -> tuple[str, dict[str, Any], str]:
+        """Extract tool call information from request data"""
+        params = request_data.get("params", {})
+        tool_name = params.get("name", "unknown")
+        tool_arguments = params.get("arguments", {})
+        request_id = request_data.get("id", "unknown")
+        return tool_name, tool_arguments, request_id
+
+    def _create_observing_streaming_response(
+        self,
+        original_response: _StreamingResponse,
+        on_complete_callback: Callable[[bytes], None],
+    ) -> StreamingResponse:
+        """Create a streaming response that yields chunks as they come and processes them after completion"""
+
+        async def observing_body_iterator():
+            body_parts: list[bytes] = []
+            try:
+                async for chunk in original_response.body_iterator:
+                    if isinstance(chunk, bytes):
+                        chunk_bytes = chunk
+                    else:
+                        chunk_bytes = str(chunk).encode("utf-8")
+
+                    # Collect chunk for later processing
+                    body_parts.append(chunk_bytes)
+
+                    # Yield chunk immediately for streaming
+                    yield chunk_bytes
+
+            except Exception as e:
+                logger.error(
```

Comment:
no needed, will already capture in sentry in the fastapi middleware

Reply:
https://github.com/WorkflowAI/WorkflowAI/pull/571/commits/d501a5832e73159064e2eb33bcfbe33bd436a29f